File: src/backtrack/51.n-queens.ts
/*
 * @lc app=leetcode id=51 lang=typescript
 *
 * [51] N-Queens
 */

// @lc code=start
function solveNQueens(n: number): string[][] {
    
    const board = Array(n).fill(0).map(() => Array(n).fill('.'));
    let res = [];

    backtrack(0);

    return res;


    function backtrack(r) {
        
        if (r === 4) {
            res.push(JSON.parse(JSON.stringify(board)));
            return;
        }

        for (let c = 0; c < 4; c++) {

            if (!isValid(r, c)) continue;

            board[r][c] = 'Q';
            backtrack(r + 1);
            board[r][c] = '.';
        }
    }

    function isValid(r, c) {
        for (let i = 0; i < n; i++) {
            if (board[i][c] === 'Q') return false;
            if (board[r][i] === 'Q') return false;

            for (let j = 0; j < n; j++) {
                if (i + j === r + c && board[i][j] === 'Q') return false;
                if (i - j === r - c && board[i][j] === 'Q') return false;
            }
        }

        return true;
    }
};

if (require.main === module) {
    let res = solveNQueens(4);

    for (const row of res) {
        console.log(row.join(' '));
    }


}
// @lc code=end

File: src/warehouse/51.n-queens.py
#
# @lc app=leetcode.cn id=51 lang=python3
# @lcpr version=30204
#
# [51] N 皇后
#


# @lcpr-template-start
from copy import deepcopy
from typing import List, Optional
from src.adt.py.leetcodeType import ListNode, TreeNode


# @lcpr-template-end
# @lc code=start
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:

        board = [[0] * n for _ in range(n)]
        res = []

        def valid(r, c):
            for i in range(n):
                if r - i >= 0 and board[r - i][c] == 1:
                    return False
                if r - i >= 0 and c - i >= 0 and board[r - i][c - i] == 1:
                    return False
                if r - i >= 0 and c + i < n and board[r - i][c + i] == 1:
                    return False

            return True

        def backtrack(row):
            if row == n:
                result = []
                for arr in board:
                    row_str = ''
                    for cell in arr:
                        if cell == 1:
                            row_str += 'Q'
                        else:
                            row_str += '.'
                    result.append(row_str)
                res.append(result)
                return

            for col in range(n):
                if not valid(row, col):
                    continue

                board[row][col] = 1
                backtrack(row + 1)
                board[row][col] = 0

        backtrack(0)

        return res


# @lc code=end


#
# @lcpr case=start
# 4\n
# @lcpr case=end

# @lcpr case=start
# 1\n
# @lcpr case=end

#
